pkg/client/cache/listwatch.go

```diff
@@ -34,22 +35,30 @@ type ListWatch struct {
 
 // ListWatch knows how to list and watch a set of apiserver resources.
 func (lw *ListWatch) List() (runtime.Object, error) {
-	return lw.Client.
-		Get().
-		Namespace(lw.Namespace).
-		Resource(lw.Resource).
-		SelectorParam("fields", lw.FieldSelector).
-		Do().
-		Get()
+	if lw.Client == nil {
```

Comment:
In practice (as the other commit mentioned) we are trying to fail faster - we have very few optional arguments, and almost all interfaces are required and so must be non-nil.


> On Jan 19, 2015, at 6:26 PM, Eric Mountain @ Amadeus <notifications@github.com> wrote:
> 
> In pkg/client/cache/listwatch.go:
> 
> > @@ -34,22 +35,30 @@ type ListWatch struct {
> >  
> >  // ListWatch knows how to list and watch a set of apiserver resources.
> >  func (lw *ListWatch) List() (runtime.Object, error) {
> > -	return lw.Client.
> > -		Get().
> > -		Namespace(lw.Namespace).
> > -		Resource(lw.Resource).
> > -		SelectorParam("fields", lw.FieldSelector).
> > -		Do().
> > -		Get()
> > +	if lw.Client == nil {
> @smarterclayton 
> Initially:
> 
> cmd/kubelet.go:main()
>  -> standalone.GetAPIServerClient(*authPath...)
>   -> clientauth.LoadFromFile(authPath)
> If LoadFromFile() returns a (nil *client.Client, err error) - which happens if e.g. the authPath refers to a file that doesn't exist, or the content doesn't unmarshall to an Info struct etc - GetAPIServerClient() returns a nil *client.Client.
> 
> This doesn't cause the kubelet to exit, rather it pursues init as normal and in particular:
> 
> cmd/kubelet.go:main()
>  -> standalone.RunKubelet(kcfg) - (kcfg.KubeClient == nil in the case above)
>   -> standalone.createAndInitKubelet(kcfg...)
>    -> kubelet.NewMainKubelet(..., kc.KubeClient...)
>     -> client.cache.NewReflector(&cache.ListWatch{kubeClient...}...).Run()
>      -> util.Forever(... r.listAndWatch() ...)
>       -> r.listAndWatch()
>        -> r.listerWatcher.List() -> panic on nil deref of r.client.Client
> And so it panics every second, and can't do it's job. So an alternative would indeed be to make the kubelet die if for whatever reason we are unable to intialise client.Config adequately. I opted for the error message because that deviated the least from the current way of working. I don't know if folks want the kubelet to exit in such a case - I guess we could instead choose to fail fast, fail early.
> 
> —
> Reply to this email directly or view it on GitHub.
>